sdk/src/logs/batch_log_processor.cc

```diff
@@ -236,45 +243,61 @@ void BatchLogProcessor::Export()
     }
     else
     {
+      std::unique_ptr<AsyncExportData> export_data(new AsyncExportData());
+      export_data->recordables =
+          nostd::span<std::unique_ptr<Recordable>>(records_arr.data(), records_arr.size());
+      auto ptr = export_data.get();
+      {
+        std::lock_guard<std::mutex> lk(synchronization_data_->async_export_data_m);
+        export_data_storage_->running_async_exports[ptr] = std::move(export_data);
+      }
+      std::weak_ptr<ExportDataStorage> export_data_watcher            = export_data_storage_;
       std::weak_ptr<SynchronizationData> synchronization_data_watcher = synchronization_data_;
-      exporter_->Export(
-          nostd::span<std::unique_ptr<Recordable>>(records_arr.data(), records_arr.size()),
-          [notify_force_flush, synchronization_data_watcher](sdk::common::ExportResult result) {
-            // TODO: Print result
-            if (synchronization_data_watcher.expired())
-            {
-              return true;
-            }
-
-            NotifyCompletion(notify_force_flush, synchronization_data_watcher.lock());
-            return true;
-          });
+      exporter_->Export(ptr->recordables, [notify_force_flush, synchronization_data_watcher, ptr,
+                                           export_data_watcher](sdk::common::ExportResult result) {
+        // TODO: Print result
+        if (synchronization_data_watcher.expired())
+        {
+          return true;
+        }
+        if (export_data_watcher.expired())
+        {
+          return true;
+        }
+        auto synchronization_data = synchronization_data_watcher.lock();
+        auto export_data          = export_data_watcher.lock();
+        {
+          std::lock_guard<std::mutex> lk(synchronization_data->async_export_data_m);
+          export_data->garbage_async_exports.emplace_back(
+              std::move(export_data->running_async_exports[ptr]));
+          export_data->running_async_exports.erase(ptr);
+        }
+
+        NotifyCompletion(notify_force_flush, synchronization_data);
+        return true;
+      });
     }
+    // wait for running async exports < max async export allowed
+    std::unique_lock<std::mutex> lock(synchronization_data_->async_export_waker_m);
+    synchronization_data_->async_export_waker.wait_for(lock, scheduled_delay_millis_, [this] {
+      std::lock_guard<std::mutex> lk(synchronization_data_->async_export_data_m);
+      return export_data_storage_->running_async_exports.size() <= max_export_async_;
+    });
+
+    // Clean up garbage exports
+    CleanUpGarbageAsyncData();
 #  endif
   } while (true);
 }
 
 #  ifdef ENABLE_ASYNC_EXPORT
-void BatchLogProcessor::WaitForShutdownCompletion()
+bool BatchLogProcessor::CleanUpGarbageAsyncData()
 {
-  // Since async export is invoked due to shutdown, need to wait
-  // for async thread to complete.
-  if (is_export_async_)
-  {
-    std::unique_lock<std::mutex> lk(synchronization_data_->async_shutdown_m);
-    while (true)
```

Comment:
Should we still need a indefinite loop to wait all running exports to finish in destructor? Or we may start to destroy a exporter when it's still running.

Reply:
Waiting logic to wait on all running exports to finish is added in Shutdown function.